### PR TITLE
Unsat cleanup

### PR DESCRIPTION
**Pull request by jleuering**

### CODE REVIEW

File: jconstraints-core/src/main/java/gov/nasa/jpf/constraints/smtlibUtility/smtsolver/SMTCMDContext.java [hunk 1]
```diff
@@ -140,24 +147,49 @@ public Result solve(Valuation result) {
...
+    Callable<String> readLineCallable = processOut::readLine;
+
+    context.echo(uid);
+    try {
+      while (!matchedUid
+          && (line = timeLimiter.callWithTimeout(readLineCallable, 500L, TimeUnit.MILLISECONDS))
+              != null) {
+        if (line.equals(uid)) {
+          matchedUid = true;
+        }
```
Comment (no-preserve-root) [inline jconstraints-core/src/main/java/gov/nasa/jpf/constraints/smtlibUtility/smtsolver/SMTCMDContext.java]:
```suggestion
        matchedUid = line.equals(uid)
```

File: jconstraints-core/src/main/java/gov/nasa/jpf/constraints/smtlibUtility/smtsolver/SMTCMDContext.java [hunk 2]
```diff
@@ -140,24 +147,49 @@ public Result solve(Valuation result) {
...
+        output.append(line);
+        output.append("\n");
+      }
+      if (output.toString().contains("Parse Error")) {
+        throw new IllegalStateException("Parse Error");
```
Comment (no-preserve-root) [inline jconstraints-core/src/main/java/gov/nasa/jpf/constraints/smtlibUtility/smtsolver/SMTCMDContext.java]: Perhaps report the `output` in the exception to help debugging=

File: jconstraints-core/src/main/java/gov/nasa/jpf/constraints/smtlibUtility/smtsolver/SMTCMDContext.java [hunk 3]
```diff
@@ -33,18 +34,28 @@
...
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
 
 public class SMTCMDContext extends SolverContext implements StoppableSolver {
   private Process p;
   SMTLibExportGenContext ctx;
 
   protected String[] command;
 
+  protected List<Expression<Boolean>> unsatCoreLastRun = null;
```
Comment (no-preserve-root) [inline jconstraints-core/src/main/java/gov/nasa/jpf/constraints/smtlibUtility/smtsolver/SMTCMDContext.java]: The visibility in this class is odd. Some parts are private, some, package-private, and some protected? The visibilities (and class finality) should be checked.

Comment (no-preserve-root) [inline jconstraints-core/src/main/java/gov/nasa/jpf/constraints/smtlibUtility/smtsolver/SMTCMDContext.java]: Isn't `= null` redundant here?

File: jconstraints-core/src/main/java/gov/nasa/jpf/constraints/smtlibUtility/smtsolver/SMTCMDContext.java [hunk 4]
```diff
@@ -98,39 +109,35 @@ public void pop(int n) {
...
       } else {
         ctx.flush();
-        while (!solverOutput.ready() && p.isAlive()) {
-          Thread.sleep(10);
-        }
-        while (solverOutput.ready() && (line = solverOutput.readLine()) != null) {
-          output.append(line);
-          output.append("\n");
-        }
+        output += readProcessOutput(ctx, solverOutput) + "\n";
```
Comment (no-preserve-root) [inline jconstraints-core/src/main/java/gov/nasa/jpf/constraints/smtlibUtility/smtsolver/SMTCMDContext.java]: Couldn't this just assign to output and omit the initialization to an empty string?

File: jconstraints-core/src/main/java/gov/nasa/jpf/constraints/smtlibUtility/smtsolver/SMTCMDContext.java [hunk 5]
```diff
@@ -98,39 +109,35 @@ public void pop(int n) {
...
-        System.err.println("Something went wrong in the SMT Solver process");
-        System.err.println(output);
-        return Result.ERROR;
-      }
-      Result res = Result.DONT_KNOW;
-      for (String l : output.toString().split("\n")) {
-        if (l.startsWith("<stdin>")) {
-          continue;
-        }
+      output += readProcessOutput(ctx, solverOutput);
```
Comment (no-preserve-root) [inline jconstraints-core/src/main/java/gov/nasa/jpf/constraints/smtlibUtility/smtsolver/SMTCMDContext.java]: Could be combined with the first assignment?

File: jconstraints-core/src/main/java/gov/nasa/jpf/constraints/smtlibUtility/smtsolver/SMTCMDContext.java [hunk 6]
```diff
@@ -140,24 +147,49 @@ public Result solve(Valuation result) {
...
+    boolean matchedUid = false;
+    ExecutorService threadExecutor = Executors.newCachedThreadPool();
+    SimpleTimeLimiter timeLimiter = SimpleTimeLimiter.create(threadExecutor);
+    Callable<String> readLineCallable = processOut::readLine;
+
+    context.echo(uid);
+    try {
+      while (!matchedUid
+          && (line = timeLimiter.callWithTimeout(readLineCallable, 500L, TimeUnit.MILLISECONDS))
+              != null) {
```
Comment (no-preserve-root) [inline jconstraints-core/src/main/java/gov/nasa/jpf/constraints/smtlibUtility/smtsolver/SMTCMDContext.java]: The formatting is difficult to read… maybe static import MILLISECONDS?

File: jconstraints-core/src/main/java/gov/nasa/jpf/constraints/smtlibUtility/smtsolver/SMTCMDContext.java [hunk 7]
```diff
@@ -140,24 +147,49 @@ public Result solve(Valuation result) {
...
+        output.append(line);
+        output.append("\n");
+      }
+      if (output.toString().contains("Parse Error")) {
+        throw new IllegalStateException("Parse Error");
+      }
+      if (!matchedUid) {
+        throw new IllegalStateException("Incomplete Output (Did not match UID)");
+      }
+      return output.toString().replace(uid, "").trim();
```
Comment (no-preserve-root) [inline jconstraints-core/src/main/java/gov/nasa/jpf/constraints/smtlibUtility/smtsolver/SMTCMDContext.java]: The semantics here are odd. Only whitespace at the beginning and end of output are trimmed, and during one execution, a newline is immediately re-added.

File: jconstraints-core/src/main/java/gov/nasa/jpf/constraints/smtlibUtility/smtsolver/SMTCMDContext.java [hunk 8]
```diff
@@ -140,24 +147,49 @@ public Result solve(Valuation result) {
...
+  }
+
+  protected static String readProcessOutput(
+      SMTLibExportGenContext context, BufferedReader processOut)
+      throws ExecutionException, InterruptedException, TimeoutException {
+    StringBuilder output = new StringBuilder();
+    String line;
+    String uid = "a" + UUID.randomUUID();
+    boolean matchedUid = false;
+    ExecutorService threadExecutor = Executors.newCachedThreadPool();
```
Comment (no-preserve-root) [inline jconstraints-core/src/main/java/gov/nasa/jpf/constraints/smtlibUtility/smtsolver/SMTCMDContext.java]: Each invocation appears to create a Cache + Thread, which is immediately discarded afterwards. Why not a field to actually use the cache?

File: jconstraints-core/src/main/java/gov/nasa/jpf/constraints/smtlibUtility/smtsolver/SMTCMDContext.java [hunk 9]
```diff
@@ -140,24 +147,49 @@ public Result solve(Valuation result) {
...
+    SimpleTimeLimiter timeLimiter = SimpleTimeLimiter.create(threadExecutor);
+    Callable<String> readLineCallable = processOut::readLine;
+
+    context.echo(uid);
```
Comment (no-preserve-root) [inline jconstraints-core/src/main/java/gov/nasa/jpf/constraints/smtlibUtility/smtsolver/SMTCMDContext.java]: Resource Leak: if this fails, the threadExecutor is not deallocated

File: jconstraints-core/src/main/java/gov/nasa/jpf/constraints/smtlibUtility/smtsolver/SMTCMDContext.java [hunk 10]
```diff
@@ -140,24 +147,49 @@ public Result solve(Valuation result) {
...
+      while (!matchedUid
+          && (line = timeLimiter.callWithTimeout(readLineCallable, 500L, TimeUnit.MILLISECONDS))
+              != null) {
+        if (line.equals(uid)) {
+          matchedUid = true;
+        }
+        output.append(line);
+        output.append("\n");
+      }
+      if (output.toString().contains("Parse Error")) {
```
Comment (no-preserve-root) [inline jconstraints-core/src/main/java/gov/nasa/jpf/constraints/smtlibUtility/smtsolver/SMTCMDContext.java]: Output is toString()ed twice

File: jconstraints-core/src/main/java/gov/nasa/jpf/constraints/smtlibUtility/smtsolver/SMTCMDContext.java [hunk 11]
```diff
@@ -33,18 +34,28 @@
...
+import java.util.concurrent.TimeoutException;
 
 public class SMTCMDContext extends SolverContext implements StoppableSolver {
   private Process p;
   SMTLibExportGenContext ctx;
 
   protected String[] command;
 
+  protected List<Expression<Boolean>> unsatCoreLastRun = null;
+  private Expression<Boolean> currentRun;
```
Comment (no-preserve-root) [inline jconstraints-core/src/main/java/gov/nasa/jpf/constraints/smtlibUtility/smtsolver/SMTCMDContext.java:53]: This appears to be a thread safety issue. (Is the class designed to be thread safe?)

File: jconstraints-core/src/main/java/gov/nasa/jpf/constraints/smtlibUtility/smtsolver/SMTCMDContext.java [hunk 12]
```diff
@@ -140,24 +147,49 @@ public Result solve(Valuation result) {
...
       }
-      return res;
-    } catch (IOException | InterruptedException e) {
-      e.printStackTrace();
     }
-    return Result.ERROR;
+    return res;
+  }
+
+  protected static String readProcessOutput(
```
Comment (no-preserve-root) [inline jconstraints-core/src/main/java/gov/nasa/jpf/constraints/smtlibUtility/smtsolver/SMTCMDContext.java]: Perhaps this should move to a utility class?

**Review by no-preserve-root — CHANGES_REQUESTED** [review]

File: jconstraints-core/src/main/java/gov/nasa/jpf/constraints/smtlibUtility/smtsolver/SMTCMDContext.java (hunk 11 above)
Comment (mmuesly) [reply jconstraints-core/src/main/java/gov/nasa/jpf/constraints/smtlibUtility/smtsolver/SMTCMDContext.java:53]: I think in its final state it should be thread safe. Eventually, it was not in the first draft.

**Review by no-preserve-root — APPROVED** [review]
Looks good to me :+1:

**Pull request merged**